modules/reindex/src/main/java/org/elasticsearch/index/reindex/ScrollableHitSource.java

```diff
@@ -0,0 +1,364 @@
+
+    protected final ESLogger logger;
+    protected final BackoffPolicy backoffPolicy;
+    protected final ThreadPool threadPool;
+    protected final Runnable countSearchRetry;
+    protected final Consumer<Throwable> fail;
+
+    public ScrollableHitSource(ESLogger logger, BackoffPolicy backoffPolicy, ThreadPool threadPool, Runnable countSearchRetry,
+            Consumer<Throwable> fail) {
+        this.logger = logger;
+        this.backoffPolicy = backoffPolicy;
+        this.threadPool = threadPool;
+        this.countSearchRetry = countSearchRetry;
+        this.fail = fail;
+    }
+
+    public final void start(Consumer<Response> onResponse) {
+        doStart(response -> {
+           setScroll(response.getScroll());
+           logger.debug("scroll returned [{}] documents with a scroll id of [{}]", response.getHits().size(), response.getScroll());
+           onResponse.accept(response);
+        });
+    }
+    protected abstract void doStart(Consumer<? super InternalResponse> onResponse);
+
+    public final void startNextScroll(TimeValue extraKeepAlive, Consumer<Response> onResponse) {
+        doStartNextScroll(scroll.get(), extraKeepAlive, response -> {
+            setScroll(response.getScroll());
+            onResponse.accept(response);
+        });
+    }
+    protected abstract void doStartNextScroll(String scroll, TimeValue extraKeepAlive, Consumer<? super InternalResponse> onResponse);
+    
+    @Override
+    public void close() {
+        String scroll = this.scroll.get();
+        if (Strings.hasLength(scroll)) {
+            clearScroll(scroll);
+        }
+    }
+    protected abstract void clearScroll(String scroll);
+
+    final void setScroll(String scroll) {
+        this.scroll.set(scroll);
+    }
+
+    /**
+     * Publicly facing response container.
+     */
+    public static class Response {
+        private final boolean timedOut;
+        private final List<SearchFailure> failures;
+        private final long totalHits;
+        private final List<? extends Hit> hits;
+
+        public Response(boolean timedOut, List<SearchFailure> failures, long totalHits, List<? extends Hit> hits) {
+            this.timedOut = timedOut;
+            this.failures = failures;
+            this.totalHits = totalHits;
+            this.hits = hits;
+        }
+
+        public boolean isTimedOut() {
+            return timedOut;
+        }
+
+        public final List<SearchFailure> getFailures() {
+            return failures;
+        }
+
+        public long getTotalHits() {
+            return totalHits;
+        }
+
+        /**
+         * The documents returned by the scroll.
+         */
+        public List<? extends Hit> getHits() {
+            return hits;
+        }
+    }
+
+    /**
+     * Internally facing response container. Extensions to {@linkplain ScrollableHitSource} use this instead of {@linkplain Response}.
+     */
+    public static class InternalResponse extends Response {
```

Comment:
It seemed more clever when I thought of it then it does now. The idea was to only expose the bits of the response that were important outside of implementations of `ScrollableHitSource` in `Response`. An information hiding thing. I don't think it helps a lot though, so I'm happy to trash it.